File: lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int parentVal = root.val;
        int leftVal = p.val;
        int rightVal = q.val;
        
        if(leftVal > parentVal  && rightVal > parentVal) return lowestCommonAncestor(root.right, p, q);
        if(leftVal < parentVal  && rightVal < parentVal) return lowestCommonAncestor(root.left, p, q);
        return root;
    }
}

File: lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        var tmp = root
        if(p != null && q != null && root != null){
            if(p?.`val` < root?.`val` && q?.`val` < root?.`val`){
              tmp = lowestCommonAncestor(root?.left, p, q)
            }
            if(p?.`val` > root?.`val` && q?.`val` > root?.`val`){
              tmp = lowestCommonAncestor(root?.right, p, q)
            }
        }
        return tmp
    }
}